listenbrainz/webserver/static/js/jsx/api-service.js

```diff
@@ -83,7 +88,116 @@ export default class APIService {
+    /*
+     *  Send a POST request to the ListenBrainz server to submit a listen
+     */
+
+    if (!isString(userToken)) {
+      throw new SyntaxError(`Expected usertoken string, got ${typeof userToken} instead`);
+    }
+    if (listenType !== "single" && listenType !== "playingNow" && listenType !== "import") {
+      throw new SyntaxError(`listenType can be "single", "playingNow" or "import", got ${listenType} instead`);
+    }
+
+    let struct = {
+      "listen_type": listenType,
+      "payload": payload,
+    }
+
+    let url = `${this.APIBaseURI}/submit-listens`;
+    
+    let delay = this.getRateLimitDelay();
+    // Halt execution for some time
+    await new Promise((resolve) => {
+      setTimeout(resolve, delay);
+    })
+    try {
+      const response = await fetch(url, {
+        method: 'POST',
+        headers: {
+          'Authorization': `Token ${userToken}`,
+          'Content-Type': 'application/json;charset=UTF-8',
+        },
+        body: JSON.stringify(struct),
+      })
+      this.updateRateLimitParameters(response);
+      return response.status; // Return status so that caller can handle appropriately
+    } catch {
+      // Retry if there is an network error
+      console.warn("Error, retrying in 3 sec");
+      setTimeout(() => this.submitListens(userToken, listenType, payload), 3000)
+    }
+  }
+
+  async getLatestImport(userName) {
+    /*
+     *  Send a GET request to the ListenBrainz server to get the latest import time
+     *  from previous imports for the user.
+     */
+
+    if (!isString(userName)) {
+      throw new SyntaxError(`Expected username string, got ${typeof userName} instead`);
+    }
+
+    let url = `${this.APIBaseURI}/latest-import?user_name=${userName}`;
+    const response = await fetch(url, {
+      method: 'GET',
+    });
+    this.checkStatus(response);
+    const result = await response.json();
+    return parseInt(result.latest_import);
+  }
+
+  async setLatestImport(userToken, timestamp) {
+    /*
+     * Send a POST request to the ListenBrainz server after the import is complete to
+     * update the latest import time on the server. This will make future imports stop
+     * when they reach this point of time in the listen history.
+     */
+
+
+    if (!isString(userToken)) {
+      throw new SyntaxError(`Expected usertoken string, got ${typeof userToken} instead`);
+    }
+    if (!isFinite(timestamp)) {
+      throw new SyntaxError(`Expected timestamp number, got ${typeof timestamp} instead`);
+    }
+    
+    let url = `${this.APIBaseURI}/latest-import`;
+    const response = await fetch(url, {
+      method: 'POST',
+      headers: {
+        'Authorization': `Token ${userToken}`,
+        'Content-Type': 'application/json;charset=UTF-8',
+      },
+      body: JSON.stringify({ts: parseInt(timestamp)}),
+    });
+    this.checkStatus(response);
+    return response.status; // Return true if timestamp is updated
+  }
+
+  getRateLimitDelay() {
```

Comment:
I think this function and the ratelimit variables should live in the importer and not here.